src/com/facebook/buck/apple/clang/ModuleMap.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2014-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.apple.clang;
+
+import org.stringtemplate.v4.ST;
+
+/**
+ * Generate a modulemap, with optional references to a swift generated swift header.
+ *
+ * <p>Note: currently only supports modulemaps for modules with an umbrella header, as otherwise the
+ * inferred submodule functionality does not work.
+ */
+public class ModuleMap {
+
+  public enum SwiftMode {
+    NO_SWIFT,
+    EXCLUDE_SWIFT_HEADER,
+    INCLUDE_SWIFT_HEADER
+  }
+
+  private final SwiftMode swiftMode;
+  private final String moduleName;
+
+  private String generatedModule;
+  private static final String template =
+      "module <module_name> {\n"
+          + "    umbrella header \"<module_name>.h\"\n"
```

Comment:
I think we should also be passing the umbrella header name explicitly as an input. I'm mainly thinking about clashes with existing `apple_library` which have headers with the same name as the module.

Reply:
Apple's convention is to have a master header which has the same name as the target. This is convention for regular modules/frameworks and required for umbrella frameworks  [1]. When there is a pre-existing header with the same name, I think it is expected that that is the umbrella header.

I can add an optional parameter to override it, but I don't think there's a strong usecase. It would also require us to add an extra field `umbrella_header` on `apple_library` to get user input for it, so maybe we should move the implementation to when that is needed?

[1] https://developer.apple.com/library/content/documentation/MacOSX/Conceptual/BPFrameworks/Tasks/IncludingFrameworks.html